### PR TITLE
Feature/issue 244 business logic

### PR DESCRIPTION
**Pull request by rasmus-storjohann-PG**

### CODE REVIEW

File: src/selectors/__tests__/tasks.test.ts [hunk 1]
```diff
@@ -88,58 +90,130 @@ describe('tasks selector', () => {
...
+            incompleteTask = new TaskBuilder().withTaxonomyTerm(aTaxonomyTerm).withCompleted(false).build();
+            incompleteTasks = { [incompleteTask.id]: incompleteTask };
         });
 
-        it('includes tasks that are not saved in my plan', () => {
-            const task = new TaskBuilder().withLocaleCode(locale.code).build();
-            const noSavedTaskIds: ReadonlyArray<string> = [];
-            const result = rejectTasksWithIdsInList(noSavedTaskIds, [task]);
-            expect(result).toEqual([task]);
+        it('should include a tasks that was not recommended and is recommended', () => {
```
Comment (bensinjin) [inline src/selectors/__tests__/tasks.test.ts]: These could use a reword. For this instance maybe something like `'returns tasks that were previously not recommended but are now recommended'`.

File: src/selectors/questionnaire/pick_questionnaire.ts [hunk 2]
```diff
@@ -1,6 +1,6 @@
 import { Store } from '../../stores';
 import { toValidOrThrow, ValidQuestionnaireStore } from '../../stores/questionnaire/stores';
 
-export const pullQuestionnaire = (appStore: Store): ValidQuestionnaireStore => (
+export const pickQuestionnaire = (appStore: Store): ValidQuestionnaireStore => (
```
Comment (bensinjin) [inline src/selectors/questionnaire/pick_questionnaire.ts]: Can you clarify when you chose pick over pull?

File: src/selectors/tasks/get_newly_recommended_tasks.ts [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+import * as R from 'ramda';
+import { getRecommendedTasks } from './get_recommended_tasks';
+
+export const getNewlyRecommendedTasks =
+    (oldAnswers: AnswersMap, newAnswers: AnswersMap, tasks: TaskMap, savedTaskIds: ReadonlyArray<Id>): ReadonlyArray<Task> => (
+        R.difference(
+            getRecommendedTasks(newAnswers, tasks, savedTaskIds),
+            getRecommendedTasks(oldAnswers, tasks, savedTaskIds),
+        )
+    );
```
Comment (bensinjin) [inline src/selectors/tasks/get_newly_recommended_tasks.ts]: Nice :)

File: src/fixtures/types/questionnaire.ts [hunk 4]
```diff
@@ -26,10 +26,18 @@ export interface AnswersMap {
     readonly [key: string]: Answer;
 }
 
+export enum State {
+    InQuestionnaire,
+    NotInQuestionnaire,
+    PopupNeeded,
+}
```
Comment (bensinjin) [inline src/fixtures/types/questionnaire.ts]: On first glance these states seem odd for a questionnaire. Maybe add some indication that you're referring to the questionnaire being viewed or the that the user is on the questionnaire route? PopupNeeded seems unrelated to the other states here and should likely be clarified further as well.

File: src/stores/questionnaire/reduce_valid_store.ts [hunk 5]
```diff
@@ -27,6 +32,28 @@ export const reduceValidStore = (store: ValidQuestionnaireStore, action?: Questi
...
+            ...store,
+            oldAnswers: store.answers,
+            state: State.InQuestionnaire,
+        });
+    }
+
+    if (store.state === State.InQuestionnaire && !isNewRouteToQuestionnaire) {
+        return new ValidQuestionnaireStore({
+            ...store,
+            state: State.PopupNeeded,
```
Comment (bensinjin) [inline src/stores/questionnaire/reduce_valid_store.ts]: So off the top of my head I think what we'd like to do is try and figure out a way to not have the `PopupNeeded` state at all and actually have the notifications store listen for the `RouteChangedAction` and then make a call to `addNotificationToStore`. We'd naturally have to tweak `modal_notification.tsx`'s props to allow a milliseconds value so you can have the modal display immediately.

File: src/fixtures/types/questionnaire.ts (hunk 4 above)
Comment (bensinjin) [reply src/fixtures/types/questionnaire.ts]: This also ties into my last comment in this PR.

File: src/stores/questionnaire/reduce_valid_store.ts (hunk 5 above)
Comment (rasmus-storjohann-PG) [reply src/stores/questionnaire/reduce_valid_store.ts]: My intuition is that this popup dialog will potentially become pretty complex so I don't think using the notification for it will be appropriate. It will cover the whole screen or almost, it will probably have functionality to tick off which tasks to add to my plan and which to leave out, etc...

File: src/fixtures/types/questionnaire.ts (hunk 4 above)
Comment (rasmus-storjohann-PG) [reply src/fixtures/types/questionnaire.ts]: I've renamed the enum values.

File: src/selectors/__tests__/tasks.test.ts (hunk 1 above)
Comment (rasmus-storjohann-PG) [reply src/selectors/__tests__/tasks.test.ts]: Fixed

File: src/components/main/main.tsx [hunk 6]
```diff
@@ -7,35 +7,46 @@ import { NotificationsConnectedComponent } from '../notifications/notifications_
...
 }
 
 export type Props = HeaderProps & FooterProps & RouterProps;
 
-export class Component extends React.Component<Props, DrawerState> {
+export interface Actions {
+    readonly routeChanged: (location: Location, action: Action) => RouteChangedAction;
+}
+
+export type MainComponentProps = Props & Actions;
```
[the remark this answers is not in the document]
Comment (rasmus-storjohann-PG) [reply src/components/main/main.tsx]: Fixed in a different PR I just put up... there may be some rebase conflicts between them, but shouldn't be too bad regardless of their merge order

**Comment by bensinjin** [comment]
@rasmus-storjohann-PG this is looking good, can you do a rebase and fix up your commit messages too?
